src/GraphQL/Types/Collections/SchemaTypes.cs

```diff
@@ -56,6 +56,46 @@ public class SchemaTypes : IEnumerable<IGraphType>
             [typeof(Uri)] = typeof(UriGraphType),
         });
 
+        static SchemaTypes()
```

Comment:
hints only here.

Reply:
I see `Preserve` also in `GraphQLBuilderBase` and `SchemaExtensions`.

Reply:
It doesn’t matter in which place to call these hints. The main thing is that the compiler when analyzing the execution flow sooner or later falls on them (and so these types will not be trimmed). Right?

Reply:
Honestly all that stuff smells. There is a feeling that we are not doing our job. Of course, if we ourselves know what types it is vital to save from triming, then our duty to save them... but the question is how to correctly draw a line for the necessary and sufficient set of hints.

Reply:
> Honestly all that stuff smells. There is a feeling that we are not doing our job. 

I really don't know what you are referring to.  If you read up on this, it is expected and normal to properly mark classes and methods with metadata when the needed types/methods are not obvious at compile-time.  This is not a design flaw, and nor is it expected that all libraries will 'just work'.

We may choose to alter our design, where possible, so less hints are necessary.  If you do not like `Preserve<T>`, which I'm guessing you are referring to, we can minimize usage of such code.  (See recent changes.)  I do admit that `Preserve<T>` feels like a hack when MS has provided us with metadata attributes to use.

Reply:
> but the question is how to correctly draw a line for the necessary and sufficient set of hints.

Not sure I follow.  If we don't provide any hints, GraphQL.NET can still be used in AOT-compiled code.  However, the user must specify every type they wish to use so the AOT compiler will not trim them.  The point of having an AOT-ready library is that it already contains the metadata so that the AOT compiler will have a better guess of what is needed than otherwise.

Ideally, our entire library would be properly marked, and there would be code warnings anywhere that might cause a runtime issue.  This is not possible since `NameFieldResolver` does not work with AOT scenarios, and `NameFieldResolver` is the default resolver unless otherwise specified for all builder methods.  Of course we could mark the builder methods as incompatible, but this would be a false positive if the user specified a field resolver with `.Resolve()`.

If we deprecate/remove `NameFieldResolver`, then we might be able to have a 100% properly annotated library.  But the fact that it is annotated does not make it 'smell'; this is normal and expected.

Reply:
>  But the fact that it is annotated does not make it 'smell'; this is normal and expected.

OK. I've read more stuff about that, for example this one https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/prepare-libraries-for-trimming?source=recommendations and now the whole idea is more clear to me. This all is just `System.Diagnostics.CodeAnalysis` stuff that guide trimmer (a part of build pipeline) to keep right things in place. Or, more accurately, eventually it guide **developer**. I feel it like nullable annotations, all that attributes are intended to produce warnings like "hey, look, this type requires public ctor, we should not trim it from resulting app". So far I'm fine having all these `DynamicallyAccessedMembers`/`DynamicDependency` annotations. What I still do not understand is the reasom to use aux `Preserve` method.